rust/lance-index/src/scalar/inverted/builder.rs

```diff
@@ -126,583 +135,427 @@ impl InvertedIndexBuilder {
         Ok(())
     }
 
     pub async fn remap(
         &mut self,
         mapping: &HashMap<u64, Option<u64>>,
+        src_store: Arc<dyn IndexStore>,
         dest_store: &dyn IndexStore,
     ) -> Result<()> {
-        // no need to remap the TokenSet,
-        // since no row_id is stored in the TokenSet
-        self.docs.remap(mapping);
-        if let Some(inverted_list) = self.inverted_list.as_ref() {
-            let schema = inverted_list_schema(self.params.with_position);
-            let mut writer = dest_store
-                .new_index_file(INVERT_LIST_FILE, schema.clone())
-                .await?;
-
-            let docs = Arc::new(self.docs.clone());
-            let batches = (0..self.tokens.next_id()).map(|token_id| {
-                let inverted_list = inverted_list.clone();
-                let docs = docs.clone();
-                async move {
-                    let batch = inverted_list
-                        .posting_batch(token_id, inverted_list.has_positions())
-                        .await?;
-                    let mut posting_builder = PostingListBuilder::from_batches(&[batch]);
-                    posting_builder.remap(mapping);
-                    let (batch, max_score) = posting_builder.to_batch(Some(docs))?;
-
-                    Result::Ok((batch, max_score))
-                }
-            });
-            let mut stream = stream::iter(batches).buffered(get_num_compute_intensive_cpus());
-            let mut offsets = Vec::new();
-            let mut max_scores = Vec::new();
-            let mut num_rows = 0;
-            while let Some((batch, max_score)) = stream.try_next().await? {
-                offsets.push(num_rows);
-                max_scores.push(max_score);
-                num_rows += batch.num_rows();
-                writer.write_record_batch(batch).await?;
-            }
-
-            let metadata = HashMap::from_iter(vec![
-                ("offsets".to_owned(), serde_json::to_string(&offsets)?),
-                ("max_scores".to_owned(), serde_json::to_string(&max_scores)?),
-            ]);
-            writer.finish_with_metadata(metadata).await?;
+        for part in self.partitions.iter() {
+            let part = InvertedPartition::load(src_store.clone(), *part).await?;
+            let mut builder = part.into_builder().await?;
+            builder.remap(mapping).await?;
+            builder.write(dest_store).await?;
         }
-
-        self.write(dest_store).await?;
+        self.write_metadata(dest_store, &self.partitions).await?;
         Ok(())
     }
 
-    #[instrument(level = "debug", skip_all)]
-    async fn write(&mut self, dest_store: &dyn IndexStore) -> Result<()> {
-        self.write_posting_lists(dest_store).await?;
-        self.write_tokens(dest_store).await?;
-        self.write_docs(dest_store).await?;
+    async fn write_metadata(&self, dest_store: &dyn IndexStore, partitions: &[u64]) -> Result<()> {
+        let metadata = HashMap::from_iter(vec![
```

Comment:
Not for this PR but maybe in the future we could just allow the index store to read & write JSON files.